// gh/lib/base.js
"use strict";
/**
 * © 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const configs = require("./configs");
// -- Config -------------------------------------------------------------------
const config = configs.getConfig();
function clone(o) {
    return JSON.parse(JSON.stringify(o));
}
exports.clone = clone;
// -- Utils --------------------------------------------------------------------
function load() { }
exports.load = load;
function find(filepath, opt_pattern) {
    return fs.readdirSync(filepath).filter(file => {
        return (opt_pattern || /.*/).test(file);
    });
}
exports.find = find;
function getUser() {
    return config.github_user || process.env.GH_USER;
}
exports.getUser = getUser;
// Export some config methods to allow plugins to access them
exports.getConfig = configs.getConfig;
exports.writeGlobalConfig = configs.writeGlobalConfig;
