// gh/lib/cmds/hello.js
"use strict";
/**
 * © 2013 Liferay, Inc. <https://liferay.com> and Node GH contributors
 * (see file: README.md)
 * SPDX-License-Identifier: BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
// -- Requires -------------------------------------------------------------------------------------
const logger = require("../logger");
const utils_1 = require("../utils");
const immer_1 = require("immer");
// -- Constants ------------------------------------------------------------------------------------
exports.name = 'Hello';
exports.DETAILS = {
    alias: 'he',
    description: 'Hello world example. Copy to start a new command.',
    commands: ['world'],
    options: {
        world: Boolean,
    },
    shorthands: {
        w: ['--world'],
    },
};
// -- Commands -------------------------------------------------------------------------------------
function run(options) {
    if (!utils_1.userRanValidFlags(exports.DETAILS.commands, options)) {
        options = immer_1.produce(options, draft => {
            draft.world = true;
        });
    }
    if (options.world) {
        world();
    }
}
exports.run = run;
function world() {
    logger.log('hello world :)');
}
